### DISCUSSION TITLE
draid vs other implementations

### DISCUSSION THREAD

**Discussion by NicholasRush** [openzfs/zfs#11620] [Ideas]
Hello everyone,

i use zfs for storing my data safe at work and at home. In the newest zfs release you have announced the new draid feature. What i want to know is, because several years before i came to zfs i used much more systems from Netapp (FAS) and they use a Raid-4 implementation. I know that this type of Raid would be very slow if writes are not cached to a nvram device and the parity disk would limit the write througput. But you get another nice features like expansion of the raid group, because all parity is stored on one or like the current implementation three drives. With a write cache ssd this would be very nice to have. Or do I misunderstand how it works here? But what makes draid better than raid-4, dp or tec?

Thanks in advance

NR

**Comment by IvanVolosyuk** [comment]
IMHO draid main selling point is very fast rebuild times, when all disks
are busy rebuilding instead of conventional raid when rebuild speed is
limited by the write performance of the spare. Also all disks including
spares are in use during normal operation, when in the conventional raid -
spares are completely idle.

**Reply by NicholasRush** [reply]
Hello, thank you very much, now I understand a little bit more about this. That is probably one of the reasons why Netapp introduced "Physical Mirror (plexes)" from raid groups at the time, so that the data is not lost during the rebuild if more than 2 or 3 hard disks break.

But a feature to mirror the entire raid group would also good thing for zfs, because then you can safely use raidz1 or raidz2 without losing the whole pool if more than two drives fail. It would be right that you can mirror the data asynchronously, but if the pool fail all applications that are run on it would also fail. That could prevent this.
But anyway with Raidz or Draid you could not expand the vdev itself.

```
# zpool status -v tank

  pool: tank
 state: ONLINE
 scrub: none requested
config:

        NAME        STATE     READ WRITE CKSUM
        tank        ONLINE       0     0     0
             plex0    ONLINE       0     0     0
               raidz2-0  ONLINE       0     0     0
                 c1t0d0  ONLINE       0     0     0
                 c2t0d0  ONLINE       0     0     0
                 c3t0d0  ONLINE       0     0     0
                 c4t0d0  ONLINE       0     0     0
                 c5t0d0  ONLINE       0     0     0
             plex1    ONLINE       0     0     0
               raidz2-0  ONLINE       0     0     0
                 c6t0d0  ONLINE       0     0     0
                 c7t0d0  ONLINE       0     0     0
                 c8t0d0  ONLINE       0     0     0
                 c9t0d0  ONLINE       0     0     0
                 c10t0d0  ONLINE       0     0     0

errors: No known data errors


```

**Reply by jammin2night** [reply]
NAME                 SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
tank                18.5T  4.02T  14.4T        -         -     0%    21%  1.00x    ONLINE  -
  draid2:2d:10c:0s  9.03T  2.09T  6.94T        -         -     0%  23.1%      -    ONLINE
    da0p1               -      -      -        -         -      -      -      -    ONLINE
    da1p1               -      -      -        -         -      -      -      -    ONLINE
    da2p1               -      -      -        -         -      -      -      -    ONLINE
    da3p1               -      -      -        -         -      -      -      -    ONLINE
    da4p1               -      -      -        -         -      -      -      -    ONLINE
    da5p1               -      -      -        -         -      -      -      -    ONLINE
    da6p1               -      -      -        -         -      -      -      -    ONLINE
    da7p1               -      -      -        -         -      -      -      -    ONLINE
    da8p1               -      -      -        -         -      -      -      -    ONLINE
    da9p1               -      -      -        -         -      -      -      -    ONLINE
  draid2:2d:10c:0s  9.03T  1.92T  7.11T        -         -     0%  21.3%      -    ONLINE
    da10p1              -      -      -        -         -      -      -      -    ONLINE
    da11p1              -      -      -        -         -      -      -      -    ONLINE
    da12p1              -      -      -        -         -      -      -      -    ONLINE
    da13p1              -      -      -        -         -      -      -      -    ONLINE
    da14p1              -      -      -        -         -      -      -      -    ONLINE
    da15p1              -      -      -        -         -      -      -      -    ONLINE
    da16p1              -      -      -        -         -      -      -      -    ONLINE
    da17p1              -      -      -        -         -      -      -      -    ONLINE
    da18p1              -      -      -        -         -      -      -      -    ONLINE
    da19p1              -      -      -        -         -      -      -      -    ONLINE
special                 -      -      -        -         -      -      -      -  -
  mirror             398G  1.46G   397G        -         -     0%  0.36%      -    ONLINE
    da22p10             -      -      -        -         -      -      -      -    ONLINE
    da21p10             -      -      -        -         -      -      -      -    ONLINE
logs                    -      -      -        -         -      -      -      -  -
  mirror            15.5G   256K  15.5G        -         -     0%  0.00%      -    ONLINE
    da22p1              -      -      -        -         -      -      -      -    ONLINE
    da21p1              -      -      -        -         -      -      -      -    ONLINE

**Reply by jammin2night** [reply]
pool: tank
 state: ONLINE
  scan: resilvered 2.93G in 00:00:47 with 0 errors on Sat Feb 19 17:28:07 2022
config:

        NAME                  STATE     READ WRITE CKSUM
        tank                  ONLINE       0     0     0
          draid2:2d:10c:0s-0  ONLINE       0     0     0
            da0p1             ONLINE       0     0     0
            da1p1             ONLINE       0     0     0
            da2p1             ONLINE       0     0     0
            da3p1             ONLINE       0     0     0
            da4p1             ONLINE       0     0     0
            da5p1             ONLINE       0     0     0
            da6p1             ONLINE       0     0     0
            da7p1             ONLINE       0     0     0
            da8p1             ONLINE       0     0     0
            da9p1             ONLINE       0     0     0
          draid2:2d:10c:0s-1  ONLINE       0     0     0
            da10p1            ONLINE       0     0     0
            da11p1            ONLINE       0     0     0
            da12p1            ONLINE       0     0     0
            da13p1            ONLINE       0     0     0
            da14p1            ONLINE       0     0     0
            da15p1            ONLINE       0     0     0
            da16p1            ONLINE       0     0     0
            da17p1            ONLINE       0     0     0
            da18p1            ONLINE       0     0     0
            da19p1            ONLINE       0     0     0
        special
          mirror-3            ONLINE       0     0     0
            da22p10           ONLINE       0     0     0
            da21p10           ONLINE       0     0     0
        logs
          mirror-2            ONLINE       0     0     0
            da22p1            ONLINE       0     0     0
            da21p1            ONLINE       0     0     0

errors: No known data errors

**Reply by IvanVolosyuk** [reply]
@jammin2night There is no distributed spares in the pool, so the rebuild time is limited by the time to copy all the data to the replacement drive, which is IMHO the main selling point of draid.
